ICRCs/ICRC-106/ICRC-106.md

```diff
@@ -0,0 +1,224 @@
+| Status |
+|:------:|
+|Draft|
+
+# ICRC-106: Standard for Associating Index Canisters with ICRC-1 Tokens
+
+## 1. Introduction
+
+Wallet applications and token management tools often need to retrieve both token metadata and transaction history for a given principal. However, identifying an associated index canister for ICRC-1 tokens is currently unstandardized, leading to inconsistencies in wallet integrations.
+
+Standard **ICRC-106** :
+1. Introduces a standard approach for indicating the presence of an index canister for ICRC-1 tokens through ledger metadata.
+2. Defines a minimal interface for the associated index canister to facilitate querying transaction history in a consistent manner.
+
+This draft standard aims to improve interoperability, simplify wallet integrations, and enable token-related applications to reliably access transaction histories.  It acts as a placeholder and documentation source until a more comprehensive standard for index canisters will be developed.
+
+
+## 2. Metadata
+
+A ledger implementing ICRC-106 MUST include the following entry in the output of the `icrc1_supported_standards` method:
+
+```candid
+record { name = "ICRC-106"; url = "https://github.com/dfinity/ICRC/blob/main/ICRCs/ICRC-106" }
+```
+
+Additionally, the ledger MUST provide the following metadata entry retrievable via the `icrc1_metadata` method:
+
+- `icrc106:index_principal` (text): The textual representation of the principal of the associated index canister.
+
+These metadata entries allow clients to discover and interact with the index canister associated with a ledger and can be retrieved using method `icrc1_metadata` defined by the ICRC-1 standard.
+
+
+Compliant ledgers MUST also implement the following endpoint for programmatically retrieving the index principal:
+
+```candid
+icrc106_get_index_principal: () -> (principal) query;
+```
+
+The metadata entry `icrc106:index_principal` and the `icrc106_get_index_principal` method MUST provide consistent information. Specifically:
+
+- The `icrc106:index_principal` metadata entry MUST represent the textual form of the principal returned by the `icrc106_get_index_principal` method.  
+- The `icrc106_get_index_principal` method MUST return the principal corresponding to the index canister associated with the ledger, as specified in the `icrc106:index_principal` metadata entry.
+
+This requirement ensures that both mechanisms reliably point to the same index canister.
+
+
+## 3. Index Canister Interface
+
+The index canister associated with the ledger SHOULD implement the following minimal Candid interface:
```

Comment:
for me it's a bit confusing that we're saying the index `SHOULD` implement the candid interface, but then further down suggest that they are definitely there and somehow part of this standard

> The index canister provides methods to facilitate querying of transaction history and metadata associated with accounts. Below is a description of the relevant methods specified in this standard, including their purpose, input, output, and typical use case.

> While the methods defined in this standard are sufficient for compliance with ICRC-106, certain implementations of the index canister may include additional methods to extend functionality. These methods are not required by ICRC-106 but may be present for advanced use cases:

> - Any changes to the index canister interface should maintain backward compatibility.
>
> By adhering to ICRC-106, ledger canisters provide a standardized mechanism for clients to discover and interact with their associated index canisters, improving integration and user experience within the Internet Computer ecosystem.

why not standardise the index canisters interface first and then how it can be discovered from a ledger?

Reply:
There's two reasons why we proceed this way:

*  standardized interfaces need to have name spaced methods, so a standardised index interface would have methods like icrc143_get_transactions(). 

* There's many different opinions about index canisters and converging on a standard will take time.  We don't want to postpone adding this metadata until we have a proper standard for index canisters. 

So this standard, which we don't plan to vote on really plays the role of documentation for: discovering the current index & the relevant part of the API of the current index.

Reply:
Ok, that makes sense. I would still try to change the wording in the current standard to avoid giving the impression there is some sort of standardised interface when in reality there is not. 

> Below is a description of the relevant methods specified in this standard [...]

> While the methods defined in this standard are sufficient for compliance with ICRC-106 [...]

> By adhering to ICRC-106, ledger canisters provide a standardized mechanism for clients to discover and interact with their associated index canisters [...]

Reply:
Maybe also make it explicit that the interface of the index is subject to further discussions and another ICRC

Reply:
Thanks for the suggestion @letmejustputthishere. I've added something along these lines in the introduction of the standard.